src/core/PhaseVocoder.cpp

```diff
@@ -0,0 +1,272 @@
+	{
+		if (!m_processedWindows[i])
+		{
+			generateWindow(i, false); // first pass, no cache
+			m_processedWindows[i] = true;
+		}
+	}
+
+	m_dataLock.unlock();
+}
+
+void PhaseVocoder::getFrames(std::vector<float>& outData, int start, int frames)
+{
+	if (m_originalBuffer.size() < 2048) { return; }
+	m_dataLock.lock();
+
+	if (typeInfo<float>::isEqual(m_scaleRatio, 1.0f))
+	{ // directly copy original data
+		std::copy_n(m_originalBuffer.data() + start, frames, outData.data());
+		m_dataLock.unlock();
+		return;
+	}
+
+	int windowMargin = s_overSampling / 2; // numbers of windows before full quality
+	int startWindow = static_cast<float>(start) / m_outStepSize - windowMargin;
+	int endWindow = static_cast<float>((start + frames)) / m_outStepSize + windowMargin;
+
+	startWindow = std::clamp(startWindow, 0, m_numWindows - 1);
+	endWindow = std::clamp(endWindow, 0, m_numWindows - 1);
+
+	// discard previous phaseSum if not processed
+	if (!m_processedWindows[startWindow])
+	{
+		std::fill_n(m_sumPhase.data() + startWindow * s_windowSize, s_windowSize, 0);
+	}
+
+	// this encompases the minimum windows needed to get full quality,
+	// which must be computed
+	for (int i = startWindow; i < endWindow; i++)
+	{
+		if (!m_processedWindows[i])
+		{
+			generateWindow(i, true); // theses should use the cache
+			m_processedWindows[i] = true;
+		}
+	}
+
+	for (int i = 0; i < frames; i++)
+	{
+		outData[i] = m_processedBuffer[start + i];
+	}
+
+	m_dataLock.unlock();
+}
+
+// adjust pv params buffers to a new scale ratio
+void PhaseVocoder::updateParams(float newRatio)
+{
+	if (m_originalBuffer.size() < 2048) { return; }
+	if (newRatio == m_scaleRatio) { return; } // nothing changed
+	m_dataLock.lock();
+
+	m_scaleRatio = newRatio;
+	m_stepSize = static_cast<float>(s_windowSize) / s_overSampling;
+	m_numWindows = static_cast<float>(m_originalBuffer.size()) / m_stepSize - s_overSampling - 1;
+	m_outStepSize = m_scaleRatio * m_stepSize; // float, else inaccurate
+	m_freqPerBin = m_originalSampleRate / s_windowSize;
+	m_expectedPhaseIn = 2. * F_PI * m_stepSize / s_windowSize;
+	m_expectedPhaseOut = 2. * F_PI * m_outStepSize / s_windowSize;
+
+	m_processedBuffer.resize(m_scaleRatio * m_originalBuffer.size(), 0);
+
+	// very slow :(
+	std::fill(m_processedWindows.begin(), m_processedWindows.end(), false);
+	std::fill(m_processedBuffer.begin(), m_processedBuffer.end(), 0);
+
+	m_dataLock.unlock();
+}
+
+// time shifts one window from originalBuffer and writes to m_processedBuffer
+// resources:
+// http://blogs.zynaptiq.com/bernsee/pitch-shifting-using-the-ft/
+// https://sethares.engr.wisc.edu/vocoders/phasevocoder.html
+// https://dsp.stackexchange.com/questions/40101/audio-time-stretching-without-pitch-shifting/40367#40367
+// https://www.guitarpitchshifter.com/
+// https://en.wikipedia.org/wiki/Window_function
+void PhaseVocoder::generateWindow(int windowNum, bool useCache)
+{
+	// declare vars
+	float real, imag, phase, magnitude, freq, deltaPhase = 0;
```

Comment:
Only `deltaPhase` is initialized to zero and the rest are left uninitialized. I'm not sure if that's intended.

Reply:
Not intended, but honestly none of the really need to be initialized. Will just remove the `= 0`

Reply:
Done